### ISSUE TITLE
Standardised Test Suite Implementations

### ISSUE THREAD

**Issue by chrisjsewell** [aiidateam/aiida-plugin-cutter#17]
So this is a bit of as amalgamation of #9 and #7, but I think its an important part of plugin development and deserves a separate issue. In essence, it would be good to have a relatively generic and endorsed implementation of a test suite that plugins can build on; which, in particular, test the end-to-end running of a calculation.

I've made some progress on this here: https://travis-ci.org/chrisjsewell/aiida-crystal17/builds/420453585. If you look at the repo/tests, you'll see that at present I'm updating the original `aiida-diff` tests, in parallel to tests for my own plugin. Then eventually the plan would be to create a PR for the [aiida-diff](https://github.com/ltalirz/aiida-diff) and remove them from my repo.

Breaking down my understanding of the required tests:

# Submission Test

At present in `aiida-diff` the `test_submit` test only calls `calc.submit`. This is a bit misleading, since the `submit` method does not call `_prepare_for_submission` (as explained in the [this tutorial](https://aiida-core.readthedocs.io/en/latest/developer_guide/devel_tutorial/code_plugin_qe.html#step-3-prepare-a-text-input)), which is the majority of the `JobCalculation` plugin code. Instead, `calc.submit` should be replaced with:

```python
    with SandboxFolder() as folder:
        subfolder, script_filename = calc.submit_test(folder=folder)
        print("inputs created successfully at {}".format(subfolder.abspath))
```

# Process Execution Test

As discussed in #7, I don't at present know how to directly retrieve the results of a submitted test. However, an alternative (and potentially more modular way) is to try directly running the `script_file` created by `submit_test`. The following code achieves this:

```python 
def test_calculation_execution(calc, allowed_returncodes=(0,)):
    """test that a calculation executes successfully"""
    from aiida.common.folders import SandboxFolder

    # output input files and scripts to temporary folder
    with SandboxFolder() as folder:

        subfolder, script_filename = calc.submit_test(folder=folder)
        print("inputs created at {}".format(subfolder.abspath))

        script_path = os.path.join(subfolder.abspath, script_filename)
        # we first need to make sure the script is executable
        st = os.stat(script_path)
        os.chmod(script_path, st.st_mode | stat.S_IEXEC)
        # now call script, NB: bash -l -c is required to access global variable loaded in .bash_profile
        returncode = subprocess.call(["bash", "-l", "-c", script_path], cwd=subfolder.abspath)

        if returncode not in allowed_returncodes:
            # the script reroutes stderr to _scheduler-stderr.txt (at least in v0.12)
            err_msg = "process failed (and couldn't find _scheduler-stderr.txt)"
            stderr_path = os.path.join(subfolder.abspath, "_scheduler-stderr.txt")
            if os.path.exists(stderr_path):
                with open(stderr_path) as f:
                    err_msg = "Process failed with stderr:\n{}".format(f.read())
            raise RuntimeError(err_msg)
        print("calculation completed execution")
```

## Testing long running executables

In reality testing that physics codes are executed would be time consuming, and also for licensed software you can't really use that on Travis. Two approaches are possible:

(1) pytest allows test to be decorated with a marker, e.g.

```python
@pytest.mark.process_execution
def test_process(new_database):
```

which can then be used to ignore certain tests, e.g. `pytest -v -m "not process_execution"` 

(2) mock executables can be created. `aiida-vasp` appears to do this, but I think I've come up with a fairly generic mock executable solution, as I explain [here](https://github.com/chrisjsewell/aiida-crystal17#testing-against-mock-crystal17-executables). The executable works by creating a hash of the input file contents, then attempting to match this to the expected (pre-computed) outputs. Switching between mock and non-mock executables, is simply a matter of setting the global variable `MOCK_EXECUTABLES=true`.

# Output Parsing and end-to-end testing

This could be achieved in the same manner as the step above, but more preferably by wrapping the calculation in a `WorkChain` and accessing the *future* of the calculation. Hopefully @sphuber  can advise

**Comment by ltalirz** [comment]
> At present in aiida-diff the test_submit test only calls calc.submit. This is a bit misleading, since the submit method does not call _prepare_for_submission

Well, it does. But I agree that calling `calc.submit_test` makes more sense.

> an alternative (and potentially more modular way) is to try directly running the script_file created by submit_test.

Thanks, I've added this for the moment.
For the new aiida version, everything becomes a process, so retrieving the results becomes as easy as `results = run(calc)` . 
Still, we should include backwards compatible test for a period of time.

> I've come up with a fairly generic mock executable solution, as I explain here.

That seems very useful to me, and we might think about including this (or at least some template of it) in the plugin cutter.

I'm going to leave this issue open until I've found time to go again through the tests & tidy up things + think about inclusion of the mock executables

**Comment by ltalirz** [comment]
The `test_calculation_execution` function is still in the codebase of the plugin cutter although not used currently. It's actually more of a `test_script_submission` function, so if a plugin provides example run scripts, one could use it to test that these actually work.
However, such scripts are likely to use `submit`, so it's anyhow not straightforward to wait until they finish.

@chrisjsewell Should we remove it or do you still see a use case for it?

**Comment by chrisjsewell** [comment]
No I don't use it anymore, just `submit_test` and `run_get_node`.
Here is the test file for my (v1.0.0b1) basic calculation: https://github.com/chrisjsewell/aiida-crystal17/blob/aiida_workshop/aiida_crystal17/tests/test_cry_basic.py

**Comment by chrisjsewell** [comment]
Note that `submit_test` is awaiting aiidateam/aiida_core#2650, but my tests pass using aiidateam/aiida_core#2658
